include/odp/api/spec/queue_types.h

```diff
@@ -122,6 +122,37 @@ typedef enum odp_nonblocking_t {
 
 } odp_nonblocking_t;
 
+/**
+ * Original event order maintenance options
+ *
+ * Options to keep or ignore the original event order of a source queue. This
+ * option is relevant for (plain or parallel scheduled) queues that are
+ * destinations for events enqueued while holding an ordered queue
+ * synchronization context. By default, an ordered context maintains original
+ * event order regardless of the destination queue type. Event re-ordering may
+ * cause extra synchronization effort for implementation and a long delay before
+ * application can receive a re-ordered event from the destination queue. This
+ * is wasteful and in some cases the extra delay is not acceptable for those
+ * destination queues that do not need to maintain the original event order.
+ * Application can use ODP_QUEUE_ORDER_IGNORE option to prevent implementation
+ * from performing unnecessary event re-ordering and negative side-effects of
+ * that.
+ */
+typedef enum odp_queue_order_t {
+	/** Keep original event order. Events enqueued into this queue while
+	 *  holding an ordered queue synchronization context maintain the
+	 *  original event order of the source queue.
+	 */
+	ODP_QUEUE_ORDER_KEEP = 0,
+
+	/** Ignore original event order. Events enqueued into this queue do not
+	 *  need to maintain the original event order of the source queue.
+	 *  Implementation must avoid significant event re-ordering delays.
```

Comment:
Must seems a bit strong here. Shouldn't this be worded similar to other performance hints? Wording such as "Implementations may ignore ordering to provide enqueue improved performance" might be more consistent with other such hints.

Reply:
Application needs to be able to trust that an enqueue operation does not wait infinitely or does not have an unexpectedly long timeout for waiting other threads. So, "must" is needed in the wording (in one form or another) instead of "may".

Reply:
Do we have scenarios where use of ordered queues can result in deadlock or livelock? As long as individual threads don't stall for other reasons they'll eventually release the ordered context, which will permit any pending reorderings to be resolved.

Reply:
Deadlock will result e.g. when application protects a plain queue with a spinlock.

lock() -> enqueue() -> unlock() ...  lock() -> while(dequeue() == EVENT_INVALID) -> unlock()

If a thread holds a lock and waits in dequeue, other threads cannot enqueue and re-order never resolves...

Reply:
Thread successfully enqueued an event into the queue, so it should be able to dequeue it from there in a finite time. Locks can be used for many reasons to protect the queue and e.g. counters calculating how many events are in the queue.